Ion 1.1 binary encoding specification (#215)
amzn/ion-docs: src/binary-encoding.adoc

= Binary Encoding
:toc:

[[encoding_primitives]]

> == Encoding primitives

Reviewer: ```suggestion
== Encoding Primitives
```

[[varuint]]

> === `VarUInt`

Reviewer 2: I think there will be confusion (and probably bugs) in spec and code if we reuse these terms across Ion versions.

Author: I'm open to renaming them. What would you suggest?

Reviewer: I don't have great suggestions, but let me suggest a few:

* `FlexInt`/`FlexUInt` - distinct spelling, only one character more.  My favorite of the bunch
* `ElasticInt`/`ElasticUInt` - a bit long in my opinion.
* `VaryInt`/`VaryUInt` - a bit close to `Var` for my tastes but _technically_ not the same if `Var` stands for *variable*.

Reviewer 3: What differentiates this from the 1.0 `Var(U)Int` is that this may be called "definite" (you know the byte count in the beginning) while 1.0 could be called "indefinite" (you only know you're done when you're done).

Reviewer 2: Those proposals all emphasize the variable-width aspect, which is fine in this context (and I kinda like "flex").  However that aspect is not distinguishing WRT fully fledged integers (or symbols), so the terms don't work well as type names in the macro-signature DSL.   There, I'm currently considering `plain_int` _etc._

Reviewer: > What differentiates this from the 1.0 Var(U)Int is that this may be called "definite" (you know the byte count in the beginning) while 1.0 could be called "indefinite" (you only know you're done when you're done).

My issue with this, is that all of the 1.1 encodings are definite.  You know the byte count for `Fixed...` integers as well.  Definite conveys something very specific about a difference of 1.0/1.1 variable length integers that I don't think stands on its own.

A variable-length unsigned integer.

The bytes of a ``VarUInt``s are written in
link:https://en.wikipedia.org/wiki/Endianness:[Little Endian byte order]. This means that the first byte will contain
the ``VarUInt``'s least significant bits.

The least significant bits in the `VarUInt` indicate the number of bytes that were used to encode the integer. The number

> of bytes can be found by counting the number of trailing zeros in the first byte(s) and adding one to it. All bits that

Reviewer 2: I think "trailing" and endian-ness evoke tension, so how about:

```suggestion
of bytes can be found by counting the number of least-significant zeros in the first byte(s) and adding one to it. All bits that
```

(and ditto below)

are more significant than the first `1` represent the magnitude of the `VarUInt`.

> ==== Example encoding of VarUInt 14 ====

Reviewer: ```suggestion
==== Example encoding of `VarUInt` for `14` ====
```

[source]
----

>               ┌──── Lowest bit is 1 (`end`), indicating

Reviewer: ```suggestion
              ┌──── Lowest bit is 1 (end), indicating
```

              │     this is the only byte.
0 0 0 1 1 1 0 1
└─────┬─────┘
unsigned int 14
----

> ==== Example encoding of VarUInt 729 ====

Reviewer: ```suggestion
==== Example encoding of `VarUInt` for `729` ====
```

[source]
----
             ┌──── There's 1 trailing zero, so this
             │     integer is two bytes wide.
            ┌┴┐
0 1 1 0 0 1 1 0  0 0 0 0 1 0 1 1
└────┬────┘      └──────┬──────┘
lowest 6 bits    highest 8 bits
of the unsigned  of the unsigned
integer          integer
----

> ==== Example encoding of VarUInt 21,043 ====

Reviewer: ```suggestion
==== Example encoding of `VarUInt` for `21,043` ====
```

└───┬───┘        └──────┬──────┘  └──────┬──────┘
lowest 6 bits    next 8 bits of   highest 8 bits
of the unsigned  the unsigned     of the unsigned
integer          integer          integer
----

NOTE: Ion 1.0 uses link:https://en.wikipedia.org/wiki/Endianness[Big Endian byte order] and indicates the width of the
integer by having
link:https://amazon-ion.github.io/ion-docs/docs/binary.html#varuint-and-varint-fields[an `end` flag bit in the highest
 bit of each byte].

[[varint]]
=== `VarInt`

A variable-length signed integer.

From an encoding perspective, ``VarInt``s are structurally similar to a `VarUInt` (<<varuint, described above>>). Both
encode their bytes using Little Endian byte order, and both use the count of trailing zero bits to indicate how many
bytes were used to encode the integer. They differ in the _interpretation_ of their bits; while a ``VarUInt``'s bits
are unsigned, a ``VarInt``'s bits are encoded using
link:https://en.wikipedia.org/wiki/Two%27s_complement[two's complement notation].

TIP: An implementation could choose to read a `VarInt` by instead reading a `VarUInt` and then reinterpreting its bits
as two's complement.

> ==== Example encoding of VarInt 14 ====

Reviewer: ```suggestion
==== Example encoding of `VarInt` for `14` ====
```

[source]
----
              ┌──── Lowest bit is 1 (`end`), indicating
              │     this is the only byte.
0 0 0 0 1 1 1 1
└─────┬─────┘

>  2's comp. 14

Reviewer 2: I think this is incorrect bits for 14

----

> ==== Example encoding of VarInt -14 ====

Reviewer: ```suggestion
==== Example encoding of `VarInt` for `-14` ====
```

[source]
----
              ┌──── Lowest bit is 1 (`end`), indicating
              │     this is the only byte.
1 1 1 0 0 1 0 1
└─────┬─────┘
 2's comp. -14
----

> ==== Example encoding of VarInt 729 ====

Reviewer: ```suggestion
==== Example encoding of `VarInt` for `729` ====
```

[source]
----
             ┌──── There's 1 trailing zero, so this
             │     integer is two bytes wide.
            ┌┴┐
0 1 1 0 0 1 1 0  0 0 0 0 1 0 1 1
└────┬────┘      └──────┬──────┘
lowest 6 bits    highest 8 bits
of the 2's       of the 2's
comp. integer    comp. integer
----

> ==== Example encoding of VarInt -729 ====

Reviewer: ```suggestion
==== Example encoding of `VarInt` for `-729` ====
```

[source]
----
             ┌──── There's 1 trailing zero, so this
             │     integer is two bytes wide.
            ┌┴┐
1 0 0 1 1 1 1 0  1 1 1 1 0 1 0 0
└────┬────┘      └──────┬──────┘
lowest 6 bits    highest 8 bits
of the 2's       of the 2's
comp. integer    comp. integer
----

[[fixeduint]]
=== `FixedUInt`

> A fixed-width, little-endian, unsigned integer whose length is inferred from the context in which it appears.

Reviewer 2: Can these be padded?

Author: As written, yes. I'm open to forbidding padded representations if you want to reserve leading zero bytes for something in the future.

Reviewer: There are no restrictions for any of our integers in Ion 1.0 and we have implementations that exploit this in varying capacity, I would not restrict it in 1.1 for any of our primitive integer encodings.

Reviewer 2: I think we should allow padding in both fixed and flex; we should be explicit here.

> ==== Example encoding of FixedUInt 3,954,261 ====

Reviewer: ```suggestion
==== Example encoding of `FixedUInt` for `3,954,261` ====
```

[source]
----

0 1 0 1 0 1 0 1  0 1 0 1 0 1 1 0  0 0 1 1 1 1 0 0
└──────┬──────┘  └──────┬──────┘  └──────┬──────┘
lowest 8 bits    next 8 bits of   highest 8 bits
of the unsigned  the unsigned     of the unsigned
integer          integer          integer
----

[[fixedint]]
=== `FixedInt`

A fixed-width, little-endian, signed integer whose length is known from the context in which it appears. Its bytes
are interpreted as two's complement.

> ==== Example encoding of FixedInt -3,954,261 ====

Reviewer: ```suggestion
==== Example encoding of `FixedInt` for `-3,954,261` ====
```

[source]
----

1 0 1 0 1 0 1 1  1 0 1 0 1 0 0 1  1 1 0 0 0 0 1 1
└──────┬──────┘  └──────┬──────┘  └──────┬──────┘
lowest 8 bits    next 8 bits of   highest 8 bits
of the 2's       the 2's comp.    of the 2's comp.
comp. integer    integer          integer
----

[[varsym]]
=== `VarSym`

> A variable-length symbol whose UTF-8 bytes can be inline, found in the symbol table, or derived from a macro

Reviewer: Should we be using *symbol token* to describe these things?

```suggestion
A variable-length symbol token whose UTF-8 bytes can be inline, found in the symbol table, or derived from a macro
```

0 0 0 1 0 1 0 1
└─────┬─────┘
  2's comp.
  positive 10
----

==== Example encoding of a `VarSym` with symbol text `hello` ====
[source]
----
              ┌─── The leading VarInt ends in a `1`,
              │    no more VarInt bytes follow.
              │      h         e        l        l        o
1 1 1 1 0 1 1 1  01101000  01100101 01101100 01101100 01101111
└─────┬─────┘    └─────────────────────┬─────────────────────┘
  2's comp.               5-byte UTF-8 encoded "hello"
  negative 5
----

==== Example encoding of `''` (symbol with empty text) using an opcode ====
[source]
----
              ┌─── The leading VarInt ends in a `1`,
              │    no more VarInt bytes follow.
              │
0 0 0 0 0 0 0 1  1110000
└─────┬─────┘    └──┬──┘
  2's comp.      opcode 0x70:
  zero           empty symbol
----

[[opcodes]]
== Opcodes

An _opcode_ is a 1-byte <<fixeduint, `FixedUInt`>> that tells the reader what the next expression represents and how the
bytes that follow should be interpreted.

=== Overview table

The meanings of each opcode are organized loosely by their high and low nibbles.

[cols="^.^1a,^.^1a,3a"]
|===
|High nibble | Low nibble | Meaning

|`0x0_` to `0x3_`

> | `*`

Reviewer: Does `*` just mean all of these are used for the encoding of that type in some way, or does it have more meaning than that?

Author: I meant it to be the Kleene star. I've replaced `*` with `A`-`F` for clarity.

> |Single-byte macro invocations

Reviewer: single-byte and multi-byte is a little strange here.  Maybe something like:

> Macro invocation where the macro ID is inline with the opcode.

|`0x4_`
| `*`

> |Multibyte macro invocations

Reviewer: Similarly maybe something like:

> Macro invocation where the macro ID follows as a `VarUInt`.

.4+|`0x5_`
| `0`-`8`

> |Integers

Reviewer 2: ```suggestion
|Fixed-width integers up to 8 bytes wide
```

Author: I changed this to:
> Integers up to 8 bytes wide

because "fixed width integer" is potentially overloaded with `int8`, `int16`, etc.

|`9`

> <|Reserved

Reviewer: I might make these things emphasized.

```suggestion
<|_Reserved_
```

|`A`-`D`
<|Floats

|`E`-`F`
<|Booleans

|`0x6_`
|`*`
|Decimals

|`0x7_`
|`*`
|Timestamps

|`0x8_`
|`*`
|Strings

|`0x9_`
|`*`
|Symbols with inline text

|`0xA_`
|`*`
|Lists

|`0xB_`
|`*`
|S-expressions

.3+|`0xC_`
|`0`
|Empty struct

|`1`

> <|Reserved

Reviewer: ```suggestion
<|_Reserved_
```

|`2`-`F`
<|Structs with symbol address field names

.2+|`0xD_`
|`0`-`1`

> |Reserved

Reviewer: ```suggestion
|_Reserved_
```

|`2`-`F`
<|Structs with `VarSym` field names

.9+|`0xE_`
|`0`
|Ion version marker

|`1`-`3`
<|Symbols with symbol table text

|`4`-`6`

> <|Annotations with symbol table text

Reviewer: We refer to this elsewhere as "*symbol address* text".  I think we should be consistent here and I tend to like the address text versus "*symbol table* text".

Reviewer 2: I think just "symbol address" is good; "symbol address text" feels self-contradictory to me.

Author: Switched to "symbol address."

|`7`-`9`
<|Annotations with `VarSym` text

|`A`
<|`null.null`

|`B`
<|Typed nulls

|`C`-`D`
<|NOP

|`E`

> <|Reserved

Reviewer: ```suggestion
<|_Reserved_
```

|`F`
<|System macro invocation

.16+|`0xF_`
|`0`
|Delimited container end

|`1`
<|Delimited list start

|`2`
<|Delimited S-expression start

|`3`
<|Delimited struct with `VarSym` field names start

|`4`
<|Variable length prefixed macro invocation

|`5`
<|Variable length integer

|`6`
<|Variable length decimal

|`7`
<|Variable length, long-form timestamp

|`8`
<|Variable length string

|`9`
<|Variable length symbol encoded as `VarSym`

|`A`
<|Variable length list

|`B`
<|Variable length S-expression

|`C`
<|Variable length struct with symbol address field names

|`D`
<|Variable length struct with `VarSym` field names

|`E`
<|Variable length blob

|`F`
<|Variable length clob

|===

[[single_byte_macro_invocations]]
=== Single-byte macro invocations

// TODO: link to macros chapter

If the value of the opcode is less than `64` (`0x40`), it represents an invocation of the macro at the corresponding
__address__—an offset within the local macro table.

==== Example encoding of a single-byte invocation of the macro at address `7`
[source]
----
0 0 0 0 0 1 1 1
└──────┬──────┘
  FixedUInt 7

> ----

Reviewer: Style note--I would model these as figures, not as sub-sections.  E.g.,

```asciidoc
.Figure {counter:figure-number}. Top-level E-expressions
[source,plain,%unbreakable]
----
a (:values 1 2 3) b (:void) c
a 1 2 3 b c
----
```

==== Example encoding of an invocation of the macro at address `31`
[source]
----
0 0 0 1 1 1 1 1
└──────┬──────┘
  FixedUInt 31
----

// TODO: Link to macro calling conventions

Note that the opcode alone tells us which macro is being invoked, but it does not supply enough information for the
reader to parse any arguments that may follow. The parsing of arguments is described in detail in the section _Macro
calling conventions_.

[[multi_byte_macro_invocations]]
=== Multibyte macro invocations

While invocations of macro addresses in the range `[0, 63]` can be encoded in a single byte using
<<single_byte_macro_invocations, single byte macro invocations>>, many applications will benefit from defining more than
64 macros.

If the high nibble of the opcode is `0x4_`, then the low nibble represents the four least significant bits of the macro

> address. A <<varuint, `VarUInt`>> follows that contains the remaining, more significant bits.

Reviewer 2: This feels backwards in endianness, when compared with the way bits carry over in VarUInt.

Author: I'm not sure I follow; could you describe the arrangement you expected?

Reviewer: My interpretation is that the least four significant bits are in the op-code, followed by a `VarUInt` that encodes the remaining bits in LE order (each byte that follows modulo the encoding hold bits from lower to higher).  So in principle the bit order is consistent, so I don't quite follow the "backwards" point as well.

Now one thing I think one could expect (though I don't know if it is worth the complexity) is that the encoding of the `VarUInt` actually starts with the low nibble of the op-code, but I think that means some special purpose machinery for this case so I am good with the definition as it stands.

Author: > My interpretation is that the least four significant bits are in the op-code, followed by a VarUInt that encodes the remaining bits in LE order (each byte that follows modulo the encoding hold bits from lower to higher). So in principle the bit order is consistent, so I don't quite follow the "backwards" point as well.

This is what I was going for. 👍 I didn't move the continuation flag bits into the opcode's low nibble because it would require special machinery.

└──┬──┘ └──┬──┘  └──────┬──────┘
   │       │            └──────────── VarUInt containing the 7 most
   │       └── 4 least significant    significant bits of the macro
opcode high    bits of the macro      address
nibble 4       address

Magnitude bits: 1000111_1011
Decoded value : 1,147
Biased value  : 1,211
----

==== Example encoding of an invocation of the macro at address `71376`
[source]
----

                              ┌─── The address VarUInt ends in `10`; the trailing
                              │    zero indicates that one more VarUInt byte follows.
                             ┌┴┐
0 1 0 0 0 0 0 0  1 0 1 0 0 1 1 0  0 1 0 0 0 1 0 1
└──┬──┘ └──┬──┘  └──────┬──────┘  └──────┬──────┘
   │       │            │                └──────────── the 8 most significant bits
   │       │            │                              of the macro address
   │       │            │
   │       │            └──────────── VarUInt containing the next 7 most
   │       └── 4 least significant    significant bits of the macro
opcode high    bits of the macro      address
nibble 4       address

Magnitude bits: 01000101_101001_0000
Decoded value : 71,312
Biased value  : 71,376
----

> NOTE: From this point on in the document, example encodings are given in hexidecimal notation.

Reviewer 2: ```suggestion
NOTE: From this point on in the document, example encodings are given in hexadecimal notation.
```

[[booleans]]
== Booleans

`0x5E` represents boolean `true`, while `0x5F` represents boolean `false`.

// XXX: Structurally, these examples should be at depth 3. However, all other example sections in the doc

> //      are at heading depth 4, which keeps them out of the TOC. These are also at depth 4 for consistency.

Reviewer: More reason why to use the figure style 😃

==== Example encoding of `true`
[source]
----
5e
----

==== Example encoding of `false`
[source]
----

> 5f

Reviewer: Super minor, I think we should standardize to upper-case for hexadecimal, we use upper elsewhere in the doc.

----

[[numbers]]
== Numbers

[[integers]]
=== Integers

Opcodes in the range `0x50` to `0x58` represent an integer. The opcode is followed by a <<fixedint, `FixedInt`>> that
represents the integer value. The low nibble of the opcode (`0x_0` to `0x_8`) indicates the size of the `FixedInt`.
Opcode `0x50` represents integer `0`; no more bytes follow.

Integers that require more than 8 bytes are encoded using the variable-length integer opcode `0xF5`,
followed by a

> <<varuint, VarUInt>> indicating how many bytes of representation data follow.

Reviewer: Should there be an example for that below?

Author: There's an example of variable-length encoding, it's just being used on a value that's only a few bytes long.

```
==== Example variable-length encoding of `-944`
[source]
----
┌──── Opcode F5 indicates a variable-length integer, VarUInt length follows
│   ┌─── VarUInt 2; a 2-byte FixedInt follows
│   │
F5 05 50 fc
      └─┬─┘
   FixedInt -944
----

==== Example encoding of `0`
[source]
----
┌──── Opcode in 50-58 range indicates integer
│┌─── Low nibble 0 indicates
││    no more bytes follow.
50
----

==== Example encoding of `17`
[source]
----
┌──── Opcode in 50-58 range indicates integer
│┌─── Low nibble 1 indicates
││    a single byte follows.
51 11
    └── FixedInt 17
----

==== Example encoding of `-944`
[source]
----
┌──── Opcode in 50-58 range indicates integer
│┌─── Low nibble 2 indicates
││    that two bytes follow.
52 50 fc
   └─┬─┘
FixedInt -944
----

==== Example variable-length encoding of `-944`
[source]
----
┌──── Opcode F5 indicates a variable-length integer, VarUInt length follows
│   ┌─── VarUInt 2; a 2-byte FixedInt follows

> │   │    that two bytes follow.

Reviewer 2: ```suggestion
│   ┌─── VarUInt 2; a 2-byte FixedInt follows.
│   │
```

F5 05 50 fc
      └─┬─┘
   FixedInt -944
----

[[floats]]
=== Floats

Float values are encoded using the IEEE-754 specification, and can be serialized in four sizes:

* 0 bits (0 bytes), representing the value 0e0 and indicated by opcode `0x5A`
* 16 bits (2 bytes, link:https://en.wikipedia.org/wiki/Half-precision_floating-point_format[half precision]),
indicated by opcode `0x5B`
* 32 bits (4 bytes, link:https://en.wikipedia.org/wiki/Single-precision_floating-point_format[single precision]),
indicated by opcode `0x5C`
* 64 bits (8 bytes, link:https://en.wikipedia.org/wiki/Double-precision_floating-point_format[double precision]),
indicated by opcode `0x5D`

Note that in the Ion data model, float values are always 64 bits. However, if a value can be losslessly serialized
in fewer than 64 bits, applications may choose to do so.

==== Example encoding of `0e0`
[source]
----
┌──── Opcode in range 5A-5D indicates a float
│┌─── Low nibble A indicates
││    a 0-length float; 0e0
5A
----

==== Example encoding of `3.14`
[source]
----
┌──── Opcode in range 5A-5D indicates a float
│┌─── Low nibble B indicates a 2-byte float
││
5B 42 47
   └─┬─┘

> half-precision 3.14

Reviewer: We have to be a little careful here with the examples, if `3.14` (I didn't check) is irrational with respect to base-2, the actual rational value has to be indicated for the smaller formats.

See: https://amazon-ion.github.io/ion-docs/docs/float.html

> In text, binary float is represented using familiar base-10 digits. While this is convenient for human representation, there is no explicit notation for expressing a particular floating point value as binary32 or binary64. Furthermore, many base-10 real numbers are irrational with respect to base-2 and cannot be expressed exactly in either binary floating point encoding (e.g. 1.1e0).
>
> Because of this asymmetry, the rules for Ion text float notation when round-tripping to Ion binary MUST be observed:
>
> * Any text notation that can be exactly represented as binary32 MAY be encoded as either binary32 or binary64 in Ion binary.
> * Any text notation that can only be exactly represented as binary64 MUST be encoded as binary64 in Ion binary.
> * Any text notation that has no exact representation (i.e. irrational in base-2 or more precision than the binary64 mantissa), MUST be encoded as binary64. This is to ensure that irrational numbers or truncated values are represented in the highest fidelity of the float data type.

Author: I had confirmed that `3.14` could be roundtripped losslessly through `f32` and `f64` using the [Float Toy](https://evanw.github.io/float-toy/) website.

Reviewer: Thanks! This was my major concern here.

----

==== Example encoding of `3.1415927`
[source]
----
┌──── Opcode in range 5A-5D indicates a float
│┌─── Low nibble C indicates a 4-byte,
││    single-precision value.
5C 40 49 0F DB
   └────┬────┘
single-precision 3.1415927
----

==== Example encoding of `3.141592653589793`
[source]
----
┌──── Opcode in range 5A-5D indicates a float
│┌─── Low nibble C indicates a 4-byte,
││    single-precision value.
5D 40 09 21 FB 54 44 2D 18
   └──────────┬──────────┘
double-precision 3.141592653589793
----

[[decimals]]
=== Decimals

If an opcode has a high nibble of `0x6_`, it represents a decimal. Low nibble values `0x_E` and below indicate
the number of trailing bytes used to encode the decimal.

The body of the decimal is encoded as a <<varint, `VarInt`>> representing its exponent, followed by a `FixedInt`
representing its coefficient. The width of the coefficient is the total length of the decimal encoding minus the length

> of the exponent. It is possible for the coefficient to have a width of zero, indicating a coefficient of `0`.

Reviewer 2: Wow, this is silly!  It would make more sense for the _exponent_ to be optional, to shave a byte from decimal-encoded integers like `17.` aka `17d0`.  Eliding the coefficient shaves bits for the far less useful set of values `0.000...`.

(Or has my brain fully decaffeinated?)

Author: > Wow, this is silly! It would make more sense for the exponent to be optional, to shave a byte from decimal-encoded integers like 17. aka 17d0. Eliding the coefficient shaves bits for the far less useful set of values 0.000....

Wow, that's a great point! I ported over the 1.0 representation without too much thought, but your approach sounds much more beneficial. I'll take a pass at writing this up.

The value `0b1111111` indicates an unknown offset. At the time of this writing (2023-05T),
link:https://en.wikipedia.org/wiki/List_of_UTC_offsets[all real-world offsets fall between `-12:00` and `+14:00`].
* *The timestamp's fractional second precision (if present) is either 3 digits (milliseconds), 6 digits (microseconds),
or 9 digits (nanoseconds).*

The following letters to are used to denote bits in each subfield in diagrams that follow. Subfields occur in the same
order in all encoding variants, and consume the same number of bits, with the exception of the fractional bits, which
consume only enough bits to represent the fractional precision supported by the opcode being used.

[cols="^1, ^1, 4"]
|===
|Letter code | Number of bits | Subfield

| *Y*
| 7
| Year

| *M*
| 4
| Month

| *D*
| 5
| Day

| *H*
| 5
| Hour

| *m*
| 6
| Minute

| *o*
| 7
| Offset

| *U*
| 1
| Unknown or UTC offset

| *s*
|6
| Second

| *f*
| 10 (ms) +
20(μs) +
30(ns) +
| Fractional second

| *-*
| n/a
| Unused
|===

==== Opcode `0x70`: Year (1 byte)
[source]
----
+=========+
|YYYY:YYY-|
+=========+
----

==== Opcode `0x71`: Month (2 bytes)
[source]
----
+=========+=========+

> |YYYY:YYYM|MMM-:----|

Reviewer 2: I observe this is big-endian layout...

Author: This is an interesting observation. Our timestamp encoding was following IEEE-754's precendent of "there are N bytes, and these bit ranges within those bytes hold significance X, Y, and Z." However, now that I think about it I would need to have the least significant time fields in the most significant bits to ensure that the field offsets for more significant time fields are always the same. I'll rework this section.

+=========+=========+
----

==== Opcode `0x72`: Day (2 bytes)
[source]
----
+=========+=========+
|YYYY:YYYM|MMMD:DDDD|
+=========+=========+
----

==== Opcode `0x73`: Hour+Minutes @ UTC or Unknown (4 bytes)

NOTE: Each encoding for a precision greater than or equal to `Hour+Minutes` comes in two flavors: one that uses a single
bit (`U`) to indicate UTC versus Unknown offset, and another that uses 7 bits (`o`) to encode the number of quarter-hours
offset from `-14:00`.

[source]
----
+=========+=========+=========+=========+
|YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmU:----|
+=========+=========+=========+=========+
----

==== Opcode `0x74`: Hour+Minutes @ Offset (5 bytes)
[source]
----
+=========+=========+=========+=========+=========+
|YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmo:oooo|oo--:----|
+=========+=========+=========+=========+=========+
----

==== Opcode `0x75`: Seconds @ UTC or Unknown (5 bytes)
[source]
----
+=========+=========+=========+=========+=========+
|YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmU:ssss|ss--:----|
+=========+=========+=========+=========+=========+
----

==== Opcode `0x76`: Seconds @ Offset (5 bytes)
[source]
----
+=========+=========+=========+=========+=========+

> |YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmo:oooo|oo--:----|

Reviewer 2: Missing the seconds subfield.

```suggestion
|YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmo:oooo|ooss:ssss|
```
I think that's right?

+=========+=========+=========+=========+=========+
----

==== Opcode `0x77`: Milliseconds @ UTC or Unknown (6 bytes)
[source]
----
+=========+=========+=========+=========+=========+=========+

> |YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmk:ssss|ssff:ffff|ffff:----|

Reviewer 2: ```suggestion
|YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmU:ssss|ssff:ffff|ffff:----|
```

+=========+=========+=========+=========+=========+=========+
----

==== Opcode `0x78`: Milliseconds @ Offset (7 bytes)
[source]
----
+=========+=========+=========+=========+=========+=========+=========+
|YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmo:oooo|ooss:ssss|ffff:ffff|ff--:----|
+=========+=========+=========+=========+=========+=========+=========+
----

==== Opcode `0x79`: Microseconds @ UTC or Unknown (7 bytes)
[source]
----
+=========+=========+=========+=========+=========+=========+=========+

> |YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmk:ssss|ssff:ffff|ffff:ffff|ffff:ff--|

Reviewer 2: ```suggestion
|YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmU:ssss|ssff:ffff|ffff:ffff|ffff:ff--|
```

+=========+=========+=========+=========+=========+=========+=========+
----

==== Opcode `0x7A`: Microseconds @ Offset (8 bytes)
[source]
----
+=========+=========+=========+=========+=========+=========+=========+=========+
|YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmo:oooo|ooss:ssss|ffff:ffff|ffff:ffff|ffff:----|
+=========+=========+=========+=========+=========+=========+=========+=========+
----

==== Opcode `0x7B`: Nanoseconds @ UTC or Unknown (8 bytes)
[source]
----
+=========+=========+=========+=========+=========+=========+=========+=========+
|YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmo:oooo|ooss:ssss|ffff:ffff|ffff:ffff|ffff:----|

> +=========+=========+=========+=========+=========+=========+=========+=========+

Reviewer 2: I think this one needs `o`s replaced with `U` and more `f`s?

It's the same `f`s as microseconds, and same in the next entry, it's late and I can't math to figure out what's wrong.

----

==== Opcode `0x7B`: Nanoseconds @ Offset (8 bytes)
[source]
----
+=========+=========+=========+=========+=========+=========+=========+=========+

> |YYYY:YYYM|MMMD:DDDD|HHHH:Hmmm|mmmo:oooo|ooss:ssss|ffff:ffff|ffff:ffff|ffff:----|

Reviewer 2: Also suspicious.

`(coefficient, exponent)` pair, similar to a <<decimals, decimal>>. However, it is illegal for the fractional
seconds value to be greater than or equal to `1.0` or less than `0.0`. For this reason, both the exponent and
the coefficient are encoded using unsigned types. The included exponent `VarUInt` is implicitly negative, preventing
the encoding of decimal numbers greater than `1.0`. The coefficient `FixedUInt` is unsigned to prevent the encoding of
fractional seconds less than `0.0`. Note that validation is still required; namely:

* An exponent value of `0` is illegal, as that would result in a fractional seconds greater than `1.0` (a whole second).
* If `coefficient * 10^-exponent > 1.0`, that `(coefficient, exponent)` pair is illegal.

If the timestamp's length is `3`, the least significant bit in the final byte (`h`) is a flag
that indicates month (`0`) or day (`1`) precision.

==== Opcode `0xF7`: Long-form timestamp
[source]
----
     1         2         3         4         5         6         7       8         n
+=========+=========+=========+=========+=========+=========+=========+=======+   +=========+

> |YYYY:YYYY|YYYY:YYMM|MMDD:DDDh|HHHH:mmmm|mmoo:oooo|oooo:ooss|ssss|----|VarUInt|...|FixedUInt|...

Reviewer 2: ```suggestion
|YYYY:YYYY|YYYY:YYMM|MMDD:DDDH|HHHH:mmmm|mmoo:oooo|oooo:ooss|ssss|----|VarUInt|...|FixedUInt|...
```

Strings longer than 15 bytes can be encoded with the `F8` opcode, which takes a <<varuint, `VarUInt`>>-encoded length
after the opcode.

==== Example encoding of the empty string (`""`)
[source]
----
┌──── Opcode in range 80-8F indicates a string
│┌─── Low nibble 0 indicates that no UTF-8 bytes follow
80
----

==== Example encoding of a 14-byte string
[source]
----
┌──── Opcode in range 80-8F indicates a string
│┌─── Low nibble E indicates that 14 UTF-8 bytes follow
││  f  o  u  r  t  e  e  n     b  y  t  e  s
8E 66 6f 75 72 74 65 65 6e 20 62 79 74 65 73
   └──────────────────┬────────────────────┘
                 UTF-8 bytes
----

==== Example encoding of a 24-byte string
[source]
----
┌──── Opcode F8 indicates a variable-length string
│  ┌─── Length: VarUInt 24
│  │   v  a  r  i  a  b  l  e     l  e  n  g  t  h     e  n  c  o  d  i  n  g
F8 31 76 61 72 69 61 62 6c 65 20 6c 65 6e 67 74 68 20 65 6e 63 6f 64 69 6e 67
      └────────────────────────────────┬────────────────────────────────────┘
                                  UTF-8 bytes
----

[[symbols_with_inline_text]]
=== Symbols with inline text

> 

Reviewer 2: Missing some preface text as with strings.

F9 31 76 61 72 69 61 62 6c 65 20 6c 65 6e 67 74 68 20 65 6e 63 6f 64 69 6e 67
      └────────────────────────────────┬────────────────────────────────────┘
                                  UTF-8 bytes
----

[[symbols_with_symbol_table_addresses]]
=== Symbols with text in the symbol table

Symbol values whose text can be found in the local symbol table are encoded using opcodes `0xE1` through `0xE3`:

* `0xE1` represents a symbol whose address in the symbol table (aka its symbol ID) is a 1-byte
<<fixeduint, `FixedUInt`>> that follows the opcode.
* `0xE2` represents a symbol whose address in the symbol table is a 2-byte <<fixeduint, `FixedUInt`>> that follows
the opcode.
* `0xE3` represents a symbol whose address in the symbol table is a <<varuint,`VarUInt`>> that follows the opcode.

Writers MUST encode a symbol address in the smallest number of bytes possible. For each opcode above, the symbol
address that is decoded is biased by the number of addresses that can be encoded in fewer bytes.

[cols="^1a,1a,1a"]
|===
|Opcode |Symbol address range |Bias

|`0xE1`
|0 to 255
|0

|`0xE2`
|256 to 16,640
|256

|`0xE3`
|16,641 to infinity

> |16,641

Reviewer 2: Doesn't two bytes give us 64k addresses, not 16k?

Author: You ever stare at `VarUInt`s so long that all bytes become 7 bits? 😵‍💫 Good catch.

│   ┌─── Length: VarUInt 24
│   │
FE 31 49 20 61 70 70 6c 61 75 64 20 79 6f 75 72 20 63 75 72 69 6f 73 69 74 79
      └────────────────────────────────┬────────────────────────────────────┘
                            24 bytes of binary data
----

[[clobs]]
=== Clobs

Opcode `FF` indicates a clob--character data of an unspecified encoding.

==== Example encoding of a 24-byte clob
[source]
----
┌──── Opcode FF indicates a clob, VarUInt length follows
│   ┌─── Length: VarUInt 24
│   │
FF 31 49 20 61 70 70 6c 61 75 64 20 79 6f 75 72 20 63 75 72 69 6f 73 69 74 79
      └────────────────────────────────┬────────────────────────────────────┘
                            24 bytes of binary data
----

[[containers]]
== Containers

Each of the container types (list, s-expression, and struct) has both a length-prefixed encoding and a delimited
encoding.

The length-prefixed encoding places more burden on the writer, but simplifies reading and enables skipping
over uninteresting values in the data stream. In contrast, the delimited encoding is simpler and faster for
writers, but requires the reader to visit each child value in turn to skip over the container.

[[lists]]

> === Lists

Reviewer 2: About here I noticed that nulls weren't always being covered in the section for each type.

Author: I considered having an entry for `null.TYPE` in each section, but figured its content would be redundant with the dedicated _Typed nulls_ section. Now that you've read the spec all the way through, would you rather have both?

Reviewer 2: Yeah, I would, so each section has full coverage of the type.

│                    ┌─── Opcode 0xF0 indicates the end of
│                    │    the most recently opened container
F2 51 01 51 02 51 03 F0
   └─┬─┘ └─┬─┘ └─┬─┘
     1     2     3
----

===== Example encoding of `(1 (2) 3)`
[source]
----
┌──── Opcode 0xF2 indicates a delimited S-expression
│        ┌─── Opcode 0xF2 begins a nested delimited S-expression
│        │        ┌─── Opcode 0xF0 closes the most recently
│        │        │    opened delimited container: the nested S-expression.
│        │        │        ┌─── Opcode 0xF0 closes the most recently opened (and still open)
│        │        │        │    delimited container: the outer S-expression.
│        │        │        │
F2 51 01 F2 51 02 F0 51 03 F0
   └─┬─┘    └─┬─┘    └─┬─┘
     1        2        3
----

[[structs]]
=== Structs

Structs have 3 available encodings:

. <<structs_with_symbol_address_field_names, Structs with symbol address field names>>
. <<structs_with_varsym_field_names, Structs with `VarSym` field names>>
. <<delimited_structs, Delimited structs with `VarSym` field names>>

[[structs_with_symbol_address_field_names]]
==== Structs with symbol address field names

> An opcode with a high nibble of `0xC_` indicates a struct with symbol address field names. The lower

Reviewer 2: Are we doing anything with the `0xC1` opcode, which would be impossible length?

Author: The opcode overview table at the top lists it as `Reserved` for now.

 │  │  ┌─────── Field name: VarUInt 10 ($10)
 │  │  │  ┌──── Opcode `F8` indicates a variable length string
 │  │  │  │  ┌─ VarUInt: 22 the string is 22 bytes long
 │  │  │  │  │  v  a  r  i  a  b  l  e     l  e  n  g  t  h     s  t  r  u  c  t
FC 33 15 F8 2D 76 61 72 69 61 62 6c 65 20 6c 65 6e 67 74 68 20 73 74 72 75 63 74
               └─────────────────────────────┬─────────────────────────────────┘
                                        UTF-8 bytes
----

[[structs_with_varsym_field_names]]
==== Structs with `VarSym` field names

NOTE: This encoding is very similar to <<structs_with_symbol_address_field_names, structs with symbol address
field names>>, but allows writers to choose between representing each field name as a symbol address
(for example: `$10`) or as inline UTF-8 bytes (for example: `"foo"`). This encoding is potentially less
dense, but offers writers significant flexibility over whether and when field names are added to the
symbol table.

An opcode with a high nibble of `0xD_` indicates a struct with <<varsym, `VarSym`>> field names. The lower
nibble of the opcode indicates how many bytes were used to encode all of its nested `(field name, value)`
pairs.

> WARNING: Empty structs MUST be written using `0xC0`. `0xD0` is a reserved opcode.

Reviewer 2: Surprising, since `[]` and `()` can be written delimited.  Why?

Author: Oops, I worded this poorly. I meant to convey that you can't use `0xD0` for an empty length-prefixed struct, because we already have `0xC0`. Since there are no fields, we can use the same opcode for both the symbol address and `VarSym` field name struct encodings. As phrased though, I accidentally forbade empty delimited struct. I'll tweak that.

If the struct's encoded byte-length is too large to be encoded in a nibble, writers may use the `0xFD` opcode
to write a variable-length struct with <<varsym, `VarSym`>> field names. The `0xFD` opcode is followed by a
<<varuint, `VarUInt`>> that indicates the byte length.

Each field in the struct is encoded as a  <<varsym, `VarSym`>> field name, followed by an
opcode-prefixed value.

===== Example encoding of `{"foo": 1, $11: 2}`
[source]
----
┌─── Opcode with high nibble `D` indicates a struct with VarSym field names
│┌── Length: 7
││ ┌─ VarSym -3      ┌─ VarSym: 11 ($11)
││ │   f  o  o       │
D9 FD 66 6F 6F 51 01 17 91 02
      └──┬───┘ └─┬─┘    └─┬─┘
      3 UTF-8    1        2
       bytes
----

> // TODO: Demonstrate splicing macro values into the struct via VarSym escape code `0x00`.

Reviewer 2: I was going to comment on this gap, maybe uncomment?

```suggestion
TODO: Demonstrate splicing macro values into the struct via VarSym escape code `0x00`.
```

[[delimited_structs]]
==== Delimited structs

Opcode `0xF3` indicates the beginning of a delimited struct with `VarSym` field names,
while opcode `0xF0` closes the most recently opened delimited container that has not yet been closed.

> NOTE: There is no delimited encoding for structs with symbol address field names.

Reviewer 2: ```suggestion
NOTE: There is no delimited encoding for structs that exclusively use symbol address field names.
```

Author: I've been using the phrase "structs with symbol address field names" to refer to the `VarUInt`-centric struct encoding. I'll try to clarify this with an extra clause.

|`0x02`
|`null.float`

|`0x03`
|`null.decimal`

|`0x04`
|`null.timestamp`

|`0x05`
|`null.string`

|`0x06`
|`null.symbol`

|`0x07`
|`null.clob`

|`0x08`
|`null.blob`

|`0x09`
|`null.list`

|`0x0A`
|`null.sexp`

|`0x0B`
|`null.struct`
|===

==== Example encoding of `null`
[source]
----
┌──── The opcode `0xEA` represents a null (null.null)
EA
----

==== Example encoding of `null.string`
[source]
----
┌──── The opcode `0xEB` indicates a typed null; a byte indicating the type follows
│  ┌──── Byte 0x05 indicates the type `string`
EB 0x5
----

[[annotations]]
=== Annotations

Annotations can be encoded either <<annotations_with_symbol_addresses, as symbol addresses>>
or <<annotations_with_varsym_text, as ``VarSym``s>>. In both encodings, the annotations sequence appears
just before the value that it decorates.

If an annotations sequence appears before one or more additional annotations sequences, the sequences
are concatenated.

It is illegal for an annotations sequence to appear before any of the following:

* The end of the stream
* A <<nops, `NOP`>>
// TODO: Links
* An E-expression (that is: a macro invocation). To add annotations to a macro invocation, see the
`annotate` macro.

> 
> If an annotations sequence appears before one or more additional annotations sequences, the sequences
> are concatenated.

Reviewer 2: Repeated paragraph.

```suggestion
```

= Binary Encoding
:toc:

[[encoding_primitives]]
== Encoding primitives

[[varuint]]
=== `VarUInt`

A variable-length unsigned integer.

The bytes of a ``VarUInt``s are written in

> link:https://en.wikipedia.org/wiki/Endianness:[Little Endian byte order]. This means that the first byte will contain

Reviewer 3: > the first byte

do you mean "first byte_s_" or can we only represent 7-byte ints (an 8 byte VarInt would use 1 byte to describe length)?

Author: Added an `s`; there's no spec-defined size limitation on `VarInt`.

> link:https://en.wikipedia.org/wiki/Endianness:[Little Endian byte order]. This means that the first byte will contain

Reviewer 4: Style note: the Wikipedia article spells this "little-endian" (hyphen, lowercase). That may be a good spelling to standardize on.

Author: Done. I think I saw the abbreviation "LE" so many times I internalized the capitalization. 😅

the ``VarUInt``'s least significant bits.

> The least significant bits in the `VarUInt` indicate the number of bytes that were used to encode the integer.

Reviewer 3: Sorry for jumping in late and missing the convo here...

I assume we considered specifying the length prefix as the "first" bits, as opposed to the least significant in the first byte(s). That strikes me as cleaner/clearer _a priori_ but I'm guessing this was chosen after some _a posteriori_ observations that it is simpler/faster to implement as specified.

Author: Good question! Consider this example from the doc:
```
             ┌──── There's 1 zero in the least significant bits, so this
             │     integer is two bytes wide.
            ┌┴┐
0 1 1 0 0 1 1 0  0 0 0 0 1 0 1 1
└────┬────┘      └──────┬──────┘
lowest 6 bits    highest 8 bits
of the unsigned  of the unsigned
integer          integer
```

When the two bytes are read into an integer, the least significant bits of the first byte will be at the "end."

```
0 0 0 0 1 0 1 1  0 1 1 0 0 1 1 0
└──────┬──────┘  └────┬────┘ └┬┘
    highest 8      lowest 6   └── Continuation flag bits
      bits           bits
```

This enables a reader to examine the first byte in the serialized form and use a "count trailing zeros" operation (on some platforms this is "reverse bits" + "count leading zeros") to get a count of the bytes and read them into an integer. Then the entire integer can be shifted right by `num_bytes` (the number of flag bits) to get the value itself.

If instead we were to store the continuation flags in the "first" (highest) bits of the first byte(s) like this:
```
 ┌──── Continuation flag bits
┌┴┐
1 0 0 1 1 0 0 1  0 0 0 0 1 0 1 1
    └────┬────┘  └──────┬──────┘
lowest 6 bits    highest 8 bits
of the unsigned  of the unsigned
integer          integer
```
that would mean that the continuation bits would appear midway through the magnitude bits like this:
```
                  ┌──── Continuation flag bits
                 ┌┴┐
0 0 0 0 1 0 1 1  0 1 1 0 0 1 1 0
└──────┬──────┘      └────┬────┘ 
    highest 8          lowest 6   
      bits               bits
```

Reviewer 3: Makes sense. Thanks!

              │     this is the only byte.
0 0 0 1 1 1 0 1
└─────┬─────┘
unsigned int 14
----

==== Example encoding of VarUInt 729 ====
[source]
----
             ┌──── There's 1 zero in the least significant bits, so this
             │     integer is two bytes wide.
            ┌┴┐
0 1 1 0 0 1 1 0  0 0 0 0 1 0 1 1
└────┬────┘      └──────┬──────┘
lowest 6 bits    highest 8 bits
of the unsigned  of the unsigned
integer          integer
----

==== Example encoding of VarUInt 21,043 ====
[source]
----
            ┌───── There are 2 zeros in the least significant bits, so this
            │      integer is three bytes wide.
          ┌─┴─┐
1 0 0 1 1 1 0 0  1 0 0 1 0 0 0 1  0 0 0 0 0 0 1 0
└───┬───┘        └──────┬──────┘  └──────┬──────┘
lowest 6 bits    next 8 bits of   highest 8 bits
of the unsigned  the unsigned     of the unsigned
integer          integer          integer
----

NOTE: Ion 1.0 uses link:https://en.wikipedia.org/wiki/Endianness[Big Endian byte order] and indicates the width of the
integer by having
link:https://amazon-ion.github.io/ion-docs/docs/binary.html#varuint-and-varint-fields[an `end` flag bit in the highest

>  bit of each byte].

Reviewer 4: Is it worth pointing out that the size characteristics are exactly the same between the two, or is that better left for a separate document?

Author: Initial versions of this doc included lots of comparisons with the 1.0 encodings, but I found it to be pretty distracting from the core content. I think I'll remove this note, especially since we're going to rename `VarUInt`/`VarInt` in this version.

It would be good to have a different document that explained our experiences with the 1.0 encodings and our rationale for changing them.

==== Example encoding of `0d0`
[source]
----
┌──── Opcode in range 60-6F indicates a decimal
│┌─── Low nibble 0 indicates a zero-byte
││    decimal; 0d0
60
----

==== Example encoding of `1.27`
[source]
----
┌──── Opcode in range 60-6F indicates a decimal
│┌─── Low nibble 3 indicates a 3-byte decimal
││
63 FD 01 FE
   └─┬─┘ └─── Exponent: 1-byte FixedInt -2
     └────── Coefficient: VarInt 127
----

==== Example variable-length encoding of `1.27`
[source]
----
┌──── Opcode F6 indicates a variable-length decimal
│
F6 07 FD 01 FE
   │  └─┬─┘ └─── Exponent: 1-byte FixedInt -2
   │    └────── Coefficient: VarInt 127
   └───────── Decimal length: VarUInt 3
----

==== Example variable-length encoding of `-0d3`
[source]
----
┌──── Opcode 6F indicates a variable-length decimal with a coefficient of -0
│
6F 03 03
   │  └────── Exponent: FixedInt 3
   └───────── Decimal length: VarUInt 1
----

[[timestamps]]
== Timestamps

NOTE: In Ion 1.0, text timestamp fields were encoded using the local time while binary timestamp fields were encoded
using UTC time. This required applications to perform conversion logic when transcribing from one format to the other.

> *In Ion 1.1, all binary timestamp fields are encoded in local time.*

Reviewer: I would add some "as per ISO 8601" or whatever.

order in all encoding variants, and consume the same number of bits, with the exception of the fractional bits, which
consume only enough bits to represent the fractional precision supported by the opcode being used.

[cols="^1, ^1, 4"]
|===
|Letter code | Number of bits | Subfield

| *Y*
| 7
| Year

| *M*
| 4
| Month

| *D*
| 5
| Day

| *H*
| 5
| Hour

| *m*
| 6
| Minute

| *o*
| 7
| Offset

| *U*
| 1
| Unknown or UTC offset

| *s*
|6
| Second

| *f*
| 10 (ms) +
20(μs) +
30(ns) +
| Fractional second

| *-*
| n/a
| Unused
|===

NOTE: Timestamps are encoded in Little Endian byte order. In the diagrams below, the bytes are read from left to right
(least significant to most significant), while the bits within each byte should be read from right to left. (Again
least significant to most significant.) +
While this encoding may complicate the diagrams somewhat, it guarantees that the timestamp's subfields (`year`, `month`,
etc.) occupy the same bit indexes regardless of how many bytes there are overall. (The last subfield,
`fractional_seconds`, always begins at the same bit index when present, but can vary in length according to the

> precision.)

Reviewer: I like this, though it is interesting that the least significant bits in our representation correspond to the most significant *field*.  So are timestamps have effective big-endian with respect to its fields.  I think this effectively makes timestamps mixed-endian.

Reviewer: Added #220.

|MYYY:YYYY|DDDD:DMMM|mmmH:HHHH|oooo:ommm|ssss:ssoo|ffff:ffff|ffff:ffff|ffff:ffff|--ff:ffff|
+=========+=========+=========+=========+=========+=========+=========+=========+=========+
----

WARNING: Opcodes `0x7D`, `0x7E`, and `7F` are illegal; they are reserved for future use.

[[long_form_timestamp]]
=== Long-form timestamp

Unlike the <<short_form_timestamp, Short-form timestamp encoding>>, which is limited to encoding
timestamps in the most commonly referenced timestamp ranges and precisions for which it optimizes,
the long-form timestamp encoding is capable of representing any valid timestamp.

The long form begins with opcode `0xF7`. A <<varuint, `VarUInt`>> follows indicating the number
of bytes that were needed to represent the timestamp. The encoding consumes the minimum number
of bytes required to represent the timestamp. The declared length can be mapped to the timestamp’s
precision as follows:

[cols="^1, 6"]
|===
|Length | Corresponding precision

| 0
| illegal

| 1
| illegal

| 2
| Year

| 3
| Month or Day (see below)

| 4
| Illegal. The hour cannot be specified without also specifying minutes.

| 5

> | Illegal

Reviewer: Suggestion to make case consistent.

```suggestion
| 0
| _Illegal_.

| 1
| _Illegal_.

| 2
| Year

| 3
| Month or Day (see below)

| 4
| _Illegal_. The hour cannot be specified without also specifying minutes.

| 5
| _Illegal_.
```

│┌─── Low nibble E indicates that 14 UTF-8 bytes follow
││  f  o  u  r  t  e  e  n     b  y  t  e  s
9E 66 6f 75 72 74 65 65 6e 20 62 79 74 65 73
   └──────────────────┬────────────────────┘
                 UTF-8 bytes
----

==== Example encoding of a symbol with 24 bytes of inline text
[source]
----
┌──── Opcode F9 indicates a variable-length symbol with inline text
│  ┌─── Length: VarUInt 24
│  │   v  a  r  i  a  b  l  e     l  e  n  g  t  h     e  n  c  o  d  i  n  g
F9 31 76 61 72 69 61 62 6c 65 20 6c 65 6e 67 74 68 20 65 6e 63 6f 64 69 6e 67
      └────────────────────────────────┬────────────────────────────────────┘
                                  UTF-8 bytes
----

[[symbols_with_symbol_table_addresses]]
=== Symbols with text in the symbol table

Symbol values whose text can be found in the local symbol table are encoded using opcodes `0xE1` through `0xE3`:

* `0xE1` represents a symbol whose address in the symbol table (aka its symbol ID) is a 1-byte
<<fixeduint, `FixedUInt`>> that follows the opcode.
* `0xE2` represents a symbol whose address in the symbol table is a 2-byte <<fixeduint, `FixedUInt`>> that follows
the opcode.

> * `0xE3` represents a symbol whose address in the symbol table is a <<varuint,`VarUInt`>> that follows the opcode.

Reviewer: Any particular reason why we special cased only up to 2-byte integers?  If we had the space  in the opcodes, 3-4 byte fixed symbols would probably be enough for all practical use cases (definitely 3), and save the variable length ones for ridiculously sized symbols (i.e., >4-5 bytes).

Author: > Any particular reason why we special cased only up to 2-byte integers? If we had the space in the opcodes, 3-4 byte fixed symbols would probably be enough for all practical use cases (definitely 3)

2 bytes gets us 65k+ addresses, which covers most use cases. I agree that 3 bytes would cover all practical use cases. There's one more `Reserved` slot in the `0xE_` nibble if we'd like to use it for that.

Author: One other thought: because the symbol addresses are biased according to their representation, the `FlexUInt`  encoding starts at 65k+. The 3-byte `FlexUInt` encoding:
```
E3 FE FF
```
would represent 16383, which when biased produces symbol address `$82175`.

A 4-byte `FlexUInt` encoding (the same size as an opcode + 3 bytes) can store up to:
```
E3 FC FF FF
```
which comes out to 2097151 before biasing and 2,162,943 after.

A dedicated 3-byte representation would hold a lot more, but I wanted to spell out the current encoding's capacity before we decide to use another opcode.

Reviewer: That's fair--I am not really eager to specify this if we think we don't need it.

Writers MUST encode a symbol address in the smallest number of bytes possible. For each opcode above, the symbol
address that is decoded is biased by the number of addresses that can be encoded in fewer bytes.

[cols="^1a,4a,4a"]
|===
|Opcode |Symbol address range |Bias

|`0xE1`
|0 to 255
|0

|`0xE2`
|256 to 65,791
|256

|`0xE3`
|65,792 to infinity
|65,792
|===

[[binary_data]]
== Binary data

[[blobs]]
=== Blobs

Opcode `FE` indicates a blob of binary data.

==== Example encoding of a 24-byte blob
[source]
----
┌──── Opcode FE indicates a blob, VarUInt length follows
│   ┌─── Length: VarUInt 24
│   │
FE 31 49 20 61 70 70 6c 61 75 64 20 79 6f 75 72 20 63 75 72 69 6f 73 69 74 79
      └────────────────────────────────┬────────────────────────────────────┘
                            24 bytes of binary data
----

[[clobs]]
=== Clobs

> Opcode `FF` indicates a clob--character data of an unspecified encoding.

Reviewer: ```suggestion
Opcode `FF` indicates a clob--binary character data of an unspecified encoding.
```

│                    ┌─── Opcode 0xF0 indicates the end of
│                    │    the most recently opened container
F2 51 01 51 02 51 03 F0
   └─┬─┘ └─┬─┘ └─┬─┘
     1     2     3
----

===== Example encoding of `(1 (2) 3)`
[source]
----
┌──── Opcode 0xF2 indicates a delimited S-expression
│        ┌─── Opcode 0xF2 begins a nested delimited S-expression
│        │        ┌─── Opcode 0xF0 closes the most recently
│        │        │    opened delimited container: the nested S-expression.
│        │        │        ┌─── Opcode 0xF0 closes the most recently opened (and still open)
│        │        │        │    delimited container: the outer S-expression.
│        │        │        │
F2 51 01 F2 51 02 F0 51 03 F0
   └─┬─┘    └─┬─┘    └─┬─┘
     1        2        3
----

[[structs]]
=== Structs

Structs have 3 available encodings:

. <<structs_with_symbol_address_field_names, Structs with symbol address field names>>
. <<structs_with_varsym_field_names, Structs with `VarSym` field names>>
. <<delimited_structs, Delimited structs with `VarSym` field names>>

[[structs_with_symbol_address_field_names]]
==== Structs with symbol address field names

> An opcode with a high nibble of `0xC_` indicates a struct with symbol address field names. The lower

Reviewer: ```suggestion
An opcode with a high nibble of `0xC_` indicates a struct with symbol address field names (which is similar to the only available encoding of structs in Ion 1.0). The lower
```

pairs.

WARNING: Empty length-prefixed structs MUST be written using `0xC0`. `0xD0` is a reserved opcode. Empty
<<delimited_structs, delimited structs>> have their own encoding.

If the struct's encoded byte-length is too large to be encoded in a nibble, writers may use the `0xFD` opcode
to write a variable-length struct with <<varsym, `VarSym`>> field names. The `0xFD` opcode is followed by a
<<varuint, `VarUInt`>> that indicates the byte length.

Each field in the struct is encoded as a  <<varsym, `VarSym`>> field name, followed by an
opcode-prefixed value.

===== Example encoding of `{"foo": 1, $11: 2}`
[source]
----
┌─── Opcode with high nibble `D` indicates a struct with VarSym field names
│┌── Length: 7
││ ┌─ VarSym -3      ┌─ VarSym: 11 ($11)
││ │   f  o  o       │
D9 FD 66 6F 6F 51 01 17 91 02
      └──┬───┘ └─┬─┘    └─┬─┘
      3 UTF-8    1        2
       bytes
----

TODO: Demonstrate splicing macro values into the struct via VarSym escape code `0x00`.

[[delimited_structs]]
==== Delimited structs

Opcode `0xF3` indicates the beginning of a delimited struct with <<varsym, `VarSym`>> field names,

> while opcode `0xF0` closes the most recently opened delimited container that has not yet been closed.

Reviewer: I think the generic closing text needs to be changed to illustrate the `VarSym` requirements.

```suggestion
Opcode `0xF3` indicates the beginning of a delimited struct with <<varsym, `VarSym`>> field names,
while the `VarSym` control sequence `0x00 0xF0` closes the current struct.
```

Author: Good catch!

delimited structs always use `VarSym`-encoded field names.

===== Example encoding of the empty struct (`{}`)
[source]
----
┌─── Opcode 0xF3 indicates the beginning of a delimited struct with `VarSym` field names.│
│  ┌─── VarSym escape code 0x00: an opcode follows
│  │  ┌─── Opcode 0xF0 indicates the end of the most
│  │  │    recently opened delimited container
F3 00 F0
----

===== Example encoding of `{"foo": 1, $11: 2}`
[source]
----
┌─── Opcode 0xF3 indicates the beginning of a delimited struct with `VarSym` field names.
│
│  ┌─ VarSym -3      ┌─ VarSym: 11 ($11)
│  │                 │        ┌─── VarSym escape code 0x00: an opcode follows
│  │                 │        │  ┌─── Opcode 0xF0 indicates the end of the most
│  │   f  o  o       │        │  │    recently opened delimited container
F3 FD 66 6F 6F 51 01 17 91 02 00 F0
      └──┬───┘ └─┬─┘    └─┬─┘
      3 UTF-8    1        2
       bytes
----

[[nulls]]
=== Nulls

The opcode `0xEA` indicates an untyped null (that is: `null`, or its alias `null.null`).

The opcode `0xEB` indicates a typed null; a byte follows whose value indicates the type. The byte-to-type
mapping is as follows:

> 

Reviewer: I wonder if we should call this the "constant" opcode and only happen to have nulls here.  Also I would indicate that "all octets that are not in this table are illegal and reserved for future use."

[cols="^1a,4a"]
|===
|Byte |Type

|`0x00`
|`null.bool`

|`0x01`
|`null.int`

|`0x02`
|`null.float`

|`0x03`
|`null.decimal`

|`0x04`
|`null.timestamp`

|`0x05`
|`null.string`

|`0x06`
|`null.symbol`

|`0x07`
|`null.clob`

|`0x08`
|`null.blob`

|`0x09`
|`null.list`

|`0x0A`
|`null.sexp`

|`0x0B`
|`null.struct`
|===

==== Example encoding of `null`
[source]
----
┌──── The opcode `0xEA` represents a null (null.null)
EA
----

==== Example encoding of `null.string`
[source]
----
┌──── The opcode `0xEB` indicates a typed null; a byte indicating the type follows
│  ┌──── Byte 0x05 indicates the type `string`

> EB 0x5

Reviewer: ```suggestion
┌──── The opcode `0xEB` indicates a typed null; a byte indicating the type follows
│  ┌──── Byte `0x05` indicates the type `string`
EB 05
```

----

[[annotations]]
=== Annotations

Annotations can be encoded either <<annotations_with_symbol_addresses, as symbol addresses>>
or <<annotations_with_varsym_text, as ``VarSym``s>>. In both encodings, the annotations sequence appears
just before the value that it decorates.

If an annotations sequence appears before one or more additional annotations sequences, the sequences
are concatenated.

It is illegal for an annotations sequence to appear before any of the following:

* The end of the stream
* A <<nops, `NOP`>>
// TODO: Links

> * An E-expression (that is: a macro invocation). To add annotations to a macro invocation, see the

Reviewer: I just realized that in the table *and* the macro invocation sections we don't call out these things as E-Expressions, I think we should be clearer about that there.

Author: Good call, done.

`annotate` macro.

[[annotations_with_symbol_addresses]]
==== Annotations with symbol addresses
Opcodes `0xE4` through `0xE6` indicate one or more annotations encoded as symbol addresses. If the opcode is:

* `0xE4`, a single <<varuint, `VarUInt`>>-encoded symbol address follows.
* `0xE5`, two <<varuint, `VarUInt`>>-encoded symbol addresses follow.
* `0xE6`, a <<varuint, `VarUInt`>> follows that represents the number of bytes needed to encode

> the annotations sequence, which can be made up of any number of `VarUInt` symbol addresses.

Reviewer: Do we have the 1.0 style fully length prefixed variant somewhere?

Author: We don't at the moment.

Reviewer: We should add a TODO somewhere to remind us about this.

Author: Added.

      └──┬───┘ └── The annotated value: `false`
      3 UTF-8
       bytes
----

==== Example encoding of `$10::foo::false`

Note that `VarSym` annotation sequences can switch between symbol address and inline text
on a per-annotation basis.

[source]
----
┌──── The opcode `0xE8` indicates two annotations encoded as VarSyms follow
│  ┌──── Annotation: VarSym 10 ($10)
│  │  ┌──── Annotation: VarSym -3; 3 bytes of UTF-8 text follow
│  │  │   f  o  o
E8 15 FD 66 6F 6F 5F
         └──┬───┘ └── The annotated value: `false`
         3 UTF-8
          bytes
----

==== Example encoding of `$10::foo::$11::false`

[source]
----
┌──── The opcode `0xE9` indicates a variable-length sequence of VarSym-encoded annotations
│  ┌──── Length: VarUInt 6
│  │  ┌──── Annotation: VarSym 10 ($10)
│  │  │  ┌──── Annotation: VarSym -3; 3 bytes of UTF-8 text follow
│  │  │  │           ┌──── Annotation: VarSym 11 ($11)
│  │  │  │   f  o  o │
E9 0D 15 FD 66 6F 6F 17 5F
            └──┬───┘    └── The annotated value: `false`
            3 UTF-8
             bytes
----

[[nops]]
=== ``NOP``s

A `NOP` (short for "no-operation") is the binary equivalent of whitespace. `NOP` bytes have no meaning,

> but can be used as padding to achieve a desired alignment.

Reviewer: I know the restrictions follow as 1.0, but it might be worth enumerating where `NOP` is allowed and how this works in the new flavors of struct.  I already saw the one error case of `NOP` after annotations--I kind of forgot if that was a 1.0 error too, so re-enumerating the validity is probably a good thing.

Author: Added.

reader to parse any arguments that may follow. The parsing of arguments is described in detail in the section _Macro
calling conventions_. (TODO: Link)

[[macro_invocation_with_address_as_varuint]]
=== Macro Invocation With Address as `VarUInt`

While invocations of macro addresses in the range `[0, 63]` can be encoded in a single byte using
<<macro_invocation_with_address_in_opcode, invocations where the address is found in the opcode>>,
many applications will benefit from defining more than 64 macros.

If the high nibble of the opcode is `0x4_`, then the low nibble represents the four least significant bits of the macro
address. A <<varuint, `VarUInt`>> follows that contains the remaining, more significant bits.

Because the first 64 macro addresses can already be encoded using high nibbles `0` to `3`, the decoded value is biased
by 64. (That is: the reader must add 64 to the decoded value. If the decoded value is `0`, the macro address that it
represents is `64`.)

Because the address is encoded using a `VarUInt`, there is no (theoretical) limit to the number of addresses that can

> be invoked. However, larger addresses require more bytes to encode. This following table shows the number of bytes

Reviewer 4: ```suggestion
be invoked. However, larger addresses require more bytes to encode. The following table shows the number of bytes
```

[source,%unbreakable]
----
┌──── Opcode in 50-58 range indicates integer
│┌─── Low nibble 2 indicates
││    that two bytes follow.
52 50 FC
   └─┬─┘
FixedInt -944
----

.Figure {counter:figure}: Encoding of integer `_-944_`
[source,%unbreakable]
----
┌──── Opcode F5 indicates a variable-length integer, VarUInt length follows
│   ┌─── VarUInt 2; a 2-byte FixedInt follows
│   │
F5 05 50 FC
      └─┬─┘
   FixedInt -944
----

[[floats]]
=== Floats

Float values are encoded using the IEEE-754 specification, and can be serialized in four sizes:

* 0 bits (0 bytes), representing the value 0e0 and indicated by opcode `0x5A`
* 16 bits (2 bytes, link:https://en.wikipedia.org/wiki/Half-precision_floating-point_format[half precision]),
indicated by opcode `0x5B`
* 32 bits (4 bytes, link:https://en.wikipedia.org/wiki/Single-precision_floating-point_format[single precision]),
indicated by opcode `0x5C`
* 64 bits (8 bytes, link:https://en.wikipedia.org/wiki/Double-precision_floating-point_format[double precision]),
indicated by opcode `0x5D`

Note that in the Ion data model, float values are always 64 bits. However, if a value can be losslessly serialized

> in fewer than 64 bits, applications may choose to do so.

Reviewer 4: Do we mean the following?

```suggestion
Note that in the Ion data model, float values are always 64 bits. However, if a value can be losslessly serialized
in fewer than 64 bits, Ion implementations may choose to do so.
```

.Figure {counter:figure}: Encoding of float `_0e0_`
[source,%unbreakable]
----
┌──── Opcode in range 5A-5D indicates a float
│┌─── Low nibble A indicates
││    a 0-length float; 0e0
5A
----

.Figure {counter:figure}: Encoding of float `_3.14e0_`
[source,%unbreakable]
----
┌──── Opcode in range 5A-5D indicates a float
│┌─── Low nibble B indicates a 2-byte float
││
5B 42 47
   └─┬─┘
half-precision 3.14
----

.Figure {counter:figure}: Encoding of float `_3.1415927e0_`
[source,%unbreakable]
----
┌──── Opcode in range 5A-5D indicates a float
│┌─── Low nibble C indicates a 4-byte,
││    single-precision value.
5C 40 49 0F DB
   └────┬────┘
single-precision 3.1415927
----

.Figure {counter:figure}: Encoding of float `_3.141592653589793e0_`
[source,%unbreakable]
----
┌──── Opcode in range 5A-5D indicates a float
│┌─── Low nibble D indicates an 8-byte,
││    double-precision value.
5D 40 09 21 FB 54 44 2D 18
   └──────────┬──────────┘
double-precision 3.141592653589793
----

[[decimals]]

> === Decimals

Reviewer 4: Let's add an example of a non-zero int-like decimal value (exponent length 0)

Author: Good idea, done.

* *14 bits for the year (`Y`)*, which is not biased.
* *12 bits for the offset*, which counts the number of minutes (not quarter-hours) from -1440
(that is: `-24:00`). An offset value of `0b111111111111` indicates an unknown offset.

If the timestamp's length is greater than or equal to `8`, it has fractional seconds that are encoded as a
`(coefficient, exponent)` pair, similar to a <<decimals, decimal>>. However, it is illegal for the fractional
seconds value to be greater than or equal to `1.0` or less than `0.0`. For this reason, both the coefficient and
the exponent are encoded using unsigned types. The included coefficient `VarUInt` is unsigned to prevent the encoding of
fractional seconds less than `0.0`. The exponent `FixedUInt` is implicitly negative, discouraging the encoding of
decimal numbers greater than `1.0`. Note that validation is still required; namely:

* An exponent value of `0` is illegal, as that would result in a fractional seconds greater than `1.0` (a whole second).
* If `coefficient * 10^-exponent > 1.0`, that `(coefficient, exponent)` pair is illegal.

If the timestamp's length is `3`, the most significant bit in the final byte (`h`) is a flag

> that indicates month (`0`) or day (`1`) precision.

Reviewer 4: ```suggestion
If the timestamp's length is `3`, the most significant bit in the final byte (`h`) is a flag
that indicates month (`0`) or day (`1`) precision. If the timestamp's length is greater than `3`, the (`h`) bit is treated as the least-significant bit of the hour (`H`) bits.
```

.Figure {counter:figure}: Delimited encoding of an empty list (`_[]_`)
[source,%unbreakable]
----
┌──── Opcode 0xF1 indicates a delimited list
│  ┌─── Opcode 0xF0 indicates the end of the most recently opened container
F1 F0
----

.Figure {counter:figure}: Delimited encoding of `_[1, 2, 3]_`
[source,%unbreakable]
----
┌──── Opcode 0xF1 indicates a delimited list
│                    ┌─── Opcode 0xF0 indicates the end of
│                    │    the most recently opened container
F1 51 01 51 02 51 03 F0
   └─┬─┘ └─┬─┘ └─┬─┘
     1     2     3
----

.Figure {counter:figure}: Delimited encoding of `_[1, [2], 3]_`
[source,%unbreakable]
----
┌──── Opcode 0xF1 indicates a delimited list
│        ┌─── Opcode 0xF1 begins a nested delimited list
│        │        ┌─── Opcode 0xF0 closes the most recently
│        │        │    opened delimited container: the nested list.
│        │        │        ┌─── Opcode 0xF0 closes the most recently opened (and still open)
│        │        │        │    delimited container: the outer list.
│        │        │        │
F1 51 01 F1 51 02 F0 51 03 F0
   └─┬─┘    └─┬─┘    └─┬─┘
     1        2        3
----

[[s_expressions]]

> === S-Expressions

Reviewer 4: Can/should we just collapse this whole section by saying, basically, that the encodings are the same as for list, but for length-prefixed `0xB_` is used and for delimited `0xF2` is used?

Author: I eliminated all of the exposition as suggested, but I've kept the example encodings. Let me know if you think I should remove the examples as well.

(for example: `$10`) or as inline UTF-8 bytes (for example: `"foo"`). This encoding is potentially less
dense, but offers writers significant flexibility over whether and when field names are added to the
symbol table.

An opcode with a high nibble of `0xD_` indicates a struct with <<varsym, `VarSym`>> field names. The lower
nibble of the opcode indicates how many bytes were used to encode all of its nested `(field name, value)`
pairs.

WARNING: Empty length-prefixed structs MUST be written using `0xC0`. `0xD0` is a reserved opcode. Empty
<<delimited_structs, delimited structs>> have their own encoding.

If the struct's encoded byte-length is too large to be encoded in a nibble, writers may use the `0xFD` opcode
to write a variable-length struct with <<varsym, `VarSym`>> field names. The `0xFD` opcode is followed by a
<<varuint, `VarUInt`>> that indicates the byte length.

Each field in the struct is encoded as a  <<varsym, `VarSym`>> field name, followed by an
opcode-prefixed value.

.Figure {counter:figure}: Length-prefixed encoding of `_{"foo": 1, $11: 2}_`
[source,%unbreakable]
----
┌─── Opcode with high nibble `D` indicates a struct with VarSym field names

> │┌── Length: 7

Reviewer 4: ```suggestion
│┌── Length: 9
```

││ ┌─ VarSym -3      ┌─ VarSym: 11 ($11)
││ │   f  o  o       │
D9 FD 66 6F 6F 51 01 17 91 02
      └──┬───┘ └─┬─┘    └─┬─┘
      3 UTF-8    1        2
       bytes
----

[sidebar]
TODO: Demonstrate splicing macro values into the struct via VarSym escape code `0x00`.

[[delimited_structs]]
==== Delimited Structs

Opcode `0xF3` indicates the beginning of a delimited struct with <<varsym, `VarSym`>> field names,
while opcode `0xF0` closes the most recently opened delimited container that has not yet been closed.

NOTE: While length-prefixed structs can choose between <<structs_with_symbol_address_field_names, structs with
symbol address field names>> and <<structs_with_varsym_field_names, structs with `VarSym` field names>>,
delimited structs always use `VarSym`-encoded field names.

.Figure {counter:figure}: Delimited encoding of the empty struct (`_{}_`)
[source,%unbreakable]
----

> ┌─── Opcode 0xF3 indicates the beginning of a delimited struct with `VarSym` field names.│

Reviewer 4: ```suggestion
┌─── Opcode 0xF3 indicates the beginning of a delimited struct with `VarSym` field names.
```

The opcode `0xEB` indicates a typed null; a byte follows whose value indicates the type. The byte-to-type
mapping is as follows:

[cols="^1a,4a"]
|===
|Byte |Type

|`0x00`
|`null.bool`

|`0x01`
|`null.int`

|`0x02`
|`null.float`

|`0x03`
|`null.decimal`

|`0x04`
|`null.timestamp`

|`0x05`
|`null.string`

|`0x06`
|`null.symbol`

|`0x07`
|`null.clob`

|`0x08`
|`null.blob`

|`0x09`
|`null.list`

|`0x0A`
|`null.sexp`

|`0x0B`
|`null.struct`
|===

.Figure {counter:figure}: Encoding of `_null_`
[source,%unbreakable]
----
┌──── The opcode `0xEA` represents a null (null.null)
EA
----

.Figure {counter:figure}: Encoding of `_null.string_`
[source,%unbreakable]
----
┌──── The opcode `0xEB` indicates a typed null; a byte indicating the type follows
│  ┌──── Byte 0x05 indicates the type `string`
EB 0x5
----

[[annotations]]
=== Annotations

Annotations can be encoded either <<annotations_with_symbol_addresses, as symbol addresses>>
or <<annotations_with_varsym_text, as ``VarSym``s>>. In both encodings, the annotations sequence appears
just before the value that it decorates.

If an annotations sequence appears before one or more additional annotations sequences, the sequences

> are concatenated.

Reviewer 4: I don't think I like this from an implementation perspective... In the reader, I'd love to be able to scan for values and record one start/end position for its annotation sequence that I can return to later if the user requests annotations. It's less simple (and quite possibly slower) if we're supporting N annotation sequences between values. Is there a killer use case I'm missing?

Author: Allowing 'chaining' means you can write annotation sequences of arbitrary length without buffering. However, in practice sequences are so rarely longer than 2 annotations (or ~4 bytes) that this probably isn't very useful. I'm ok with removing it.

If an opcode has a high nibble of `0x6_`, it represents a decimal. Low nibble values `0x_E` and below indicate
the number of trailing bytes used to encode the decimal.

The body of the decimal is encoded as a <<varint, `VarInt`>> representing its coefficient, followed by a `FixedInt`
representing its exponent. The width of the exponent is the total length of the decimal encoding minus the length
of the coefficient. It is possible for the exponent to have a width of zero, indicating an exponent of `0`.

Decimal values that require more than 14 bytes can be encoded using the variable-length decimal opcode: `0xF6`.

A decimal with a coefficient of `-0` (which cannot be encoded as a `VarInt`) is encoded using opcode `6F`.

> The opcode is followed by a `VarUInt` representing the byte length and a `FixedInt` representing the exponent.

Reviewer 2: I'm guessing this needs to change as well

Reviewer 2: Oh actually I guess it was wrong before (should've read "coefficient ... FixedInt").

But now I suggest this just be a single VarInt, no need for a length prefix.  It'll be more compact and faster!

Author: > But now I suggest this just be a single VarInt, no need for a length prefix. It'll be more compact and faster!

Sold!

60
----

.Figure {counter:figure}: Encoding of decimal `_7d0_`
[source,%unbreakable]
----
┌──── Opcode in range 60-6F indicates a decimal
│┌─── Low nibble 1 indicates a 1-byte decimal
││

> 60 0F

Reviewer 4: ```suggestion
61 0F
```

Timestamps may be encoded using the short form if they meet all of the following conditions:

The year is between 1970 and 2097.:: The year subfield is encoded as the number of years since 1970. 7 bits are
dedicated to representing the biased year, allowing timestamps through the year 2097 to be encoded in this form.
The local offset is either UTC, unknown, or falls between `-14:00` to `+14:00` and is divisible by 15 minutes.:: 7
bits are dedicated to representing the local offset as the number of quarter hours from -56 (that is: offset `-14:00`).
The value `0b1111111` indicates an unknown offset. At the time of this writing (2023-05T),
link:https://en.wikipedia.org/wiki/List_of_UTC_offsets[all real-world offsets fall between `-12:00` and `+14:00`
and are multiples of 15 minutes].
The fractional seconds are a common precision.::
The timestamp's fractional second precision (if present) is either 3 digits (milliseconds), 6 digits (microseconds),
or 9 digits (nanoseconds).

==== Opcodes by precision and offset

Each opcode with a high nibble of `0x7_` indicates a different precision and offset encoding pair.

[cols="^1a,^1a,^1a,.^4a"]
|===

> |Opcode | Precision | Serialized size in bytes{asterisk} | Offset encoding

Reviewer: This is a great table!

| Unused
|===

NOTE: Timestamps are encoded in Little Endian byte order. In the diagrams below, the bytes are read from left to right
(least significant to most significant), while the bits within each byte should be read from right to left. (Again
least significant to most significant.) +
 +
While this encoding may complicate human reading, it guarantees that the timestamp's subfields (`year`, `month`,
etc.) occupy the same bit indexes regardless of how many bytes there are overall. (The last subfield,
`fractional_seconds`, always begins at the same bit index when present, but can vary in length according to the
precision.) This allows processors to read the Little-Endian bytes into an integer and then mask the appropriate
bit ranges to access the subfields.

==== Opcode `0x70`: Year (1 byte)
[source]
----
+=========+
|-YYY:YYYY|
+=========+
----

==== Opcode `0x71`: Month (2 bytes)
[source]
----
+=========+=========+

> |MYYY:YYYY|----:-MMM|

Reviewer 2: Yeah this is what I was talking about elsewhere, this feels more consistent endian-ness.

┌──── Opcode in range 80-8F indicates a string
│┌─── Low nibble 0 indicates that no UTF-8 bytes follow
80
----

.Figure {counter:figure}: Encoding of a 14-byte string
[source,%unbreakable]
----
┌──── Opcode in range 80-8F indicates a string
│┌─── Low nibble E indicates that 14 UTF-8 bytes follow
││  f  o  u  r  t  e  e  n     b  y  t  e  s
8E 66 6F 75 72 74 65 65 6E 20 62 79 74 65 73
   └──────────────────┬────────────────────┘
                 UTF-8 bytes
----

.Figure {counter:figure}: Encoding of a 24-byte string
[source,%unbreakable]
----
┌──── Opcode F8 indicates a variable-length string
│  ┌─── Length: VarUInt 24
│  │   v  a  r  i  a  b  l  e     l  e  n  g  t  h     e  n  c  o  d  i  n  g
F8 31 76 61 72 69 61 62 6C 65 20 6C 65 6E 67 74 68 20 65 6E 63 6f 64 69 6E 67
      └────────────────────────────────┬────────────────────────────────────┘
                                  UTF-8 bytes
----

[[symbols_with_inline_text]]
=== Symbols With Inline Text

If the high nibble of the opcode is `0x9_`, it represents a symbol whose text follows the opcode. The low nibble of the
opcode indicates how many UTF-8 bytes follow. Opcode `0x90` represents a symbol with empty text (`''`).

> .Figure {counter:figure}: Encoding of a symbol with empty text (`_''_`)

Reviewer 2: We've got seemingly inconsistent typesetting conventions; I don't understand why the `''` here, and other sample data elsewhere, is italicized.

Author: The example data is being italicized in the context of figure labels. It's subjective, but I found the un-italicized data to be visually jarring:

![image](https://github.com/amazon-ion/ion-docs/assets/611616/30c1ccc3-d297-4ac3-bde1-5a1e573c92b3)

while the italicized version felt more integrated:

![image](https://github.com/amazon-ion/ion-docs/assets/611616/65ff52ab-4547-48cc-a4dd-d09974647905)

If the timestamp's length is greater than or equal to `8`, it has fractional seconds that are encoded as a
`(coefficient, exponent)` pair, similar to a <<decimals, decimal>>. However, it is illegal for the fractional
seconds value to be greater than or equal to `1.0` or less than `0.0`. For this reason, both the coefficient and
the exponent are encoded using unsigned types. The included coefficient `FlexUInt` is unsigned to prevent the encoding of
fractional seconds less than `0.0`. The exponent `FixedUInt` is implicitly negative, discouraging the encoding of

> decimal numbers greater than `1.0`. Note that validation is still required; namely:

Reviewer: What is interesting, is what we're calling an implicitly negative exponent, is what other people would call *scale* and not only that, but our restrictions on the *coefficient*/*scale* are what, in other contexts, would be called a *mantissa*.  I am not sure that we need to introduce mantissa (fractional seconds is probably good enough), but we might consider calling this negative exponent *scale*.

Author: I'm ok with renaming those, but I'd like to leave that for a future change as I'll need to read up on the terminology.

Reviewer: Added #221.

nibble of the opcode indicates how many bytes were used to encode all of its nested `(field name, value)`
pairs.

> WARNING: Empty length-prefixed structs MUST be written using `0xC0`. `0xD0` is a reserved opcode. Empty
> <<delimited_structs, delimited structs>> have their own encoding.

Reviewer 2: ```suggestion
WARNING: This form cannot be used to encode an empty struct; `0xD0` is a reserved opcode.  Empty structs can be written using either the length-prefixed form `0xC0` or the <<delimited_structs, delimited form>> `0xF3 0xF0`.
```
